enhancements/microshift/disabling-storage-components.md

```diff
@@ -0,0 +1,239 @@
+4. MicroShift attempts to dynamically create generate the LVMD config. If it cannot, an error is logged, LVMS and CSI
+installation is skipped and install continues.
+5. If all checks pass, LVMS and CSI are deployed.
+
+**_Installation without CSI Driver and Snapshotting_**
+
+1. User installs MicroShift.
+2. User specifies a MicroShift config sets fields to disable LVMS and CSI Snapshots.
+3. MicroShift starts and detects the provided config.
+4. LVMS and CSI snapshot components are not deployed.
+
+**_Post-Start Installation, LVMS only_**
+
+1. User has already installed and started MicroShift service.
+2. User later determines there is a requirement for persistent storage, but not snapshotting.
+3. User edits the MicroShift config, enabling LVMS only.
+4. User restarts MicroShift service.
+5. MicroShift starts and detects the provided config.
+6. MicroShift performs LVMS startup checks.
+7. If all checks pass, LVMS is deployed. Else if checks fail, an error is logged, LVMS deployment is skipped, 
+and startup continues.
+
+**_Complete Uninstallation, with Data removal_**
+
+> NOTE: Installation is not reversible. Deleting LVMS or the CSI snapshotter while there are still storage volumes risks
+orphaning volumes. It is the user's responsibility to manually destroy data volumes and/or snapshots before
+uninstalling.
+
+1. User has already deployed a cluster with LVMS and CSI Snapshotting installed.
+2. User has deployed cluster workloads with persistent storage.
+3. User decides that LVMS and snapshotting are no longer needed.
+4. User takes actions to back up, wipe, or otherwise ensure data will not be irretrievable.
+5. User stops workloads with mounted storage volumes.
+6. User deletes VolumeSnapshots and waits for deletion of VolumeSnapshotContent objects to verify. The deletion process
+cannot happen after the CSI Snapshotter is deleted.
+7. User delete PersistentVolumeClaims and waits for deletion of PersistentVolumes. The deletion process
+cannot happen after LVMS is deleted.
+8. User deletes the relevant LVMS and Snapshotter cluster API resources:
+    ```shell
+    $ oc delete -n kube-system deployment.apps/csi-snapshot-controller deployment.apps/csi-snapshot-webhook
+    $ oc delete -n openshift-storage daemonset.apps/topolvm-node
+    $ oc delete -n openshift-storage deployment.apps/topolvm-controller
+    $ oc delete -n openshift-storage configmaps/lvmd
+    $ oc delete storageclasses.storage.k8s.io/topolvm-provisioner
+    ```
+9. Component deletion completes. On restart, MicroShift will not deploy LVMS or the snapshotter.
+
+### API Extensions
+
+- MicroShift config will be extended from the root with a field called `storage`, which will have 2 subfields.
+  - `.storage.driver`: **ENUM**, type is preferable to leave room for future supported drivers
+    - One of ["None|none", "lvms"]
+    - Default, nil, empty: ["lvms"]
+  - `.storage.csi-snapshot:` **BOOL** 
```

Comment:
I wonder if we should use here enum as well. Just in case. Following the general trend of avoiding booleans which later are not enough.

Reply:
An enum is a little clunky for this field. Perhaps an array of CSI components?

Reply:
But wouldn't this be overkill? I think micro-lvms merged everything back into one image so the only valid array item would be `csi-snapshot`?

Reply:
We tend to favor string-like fields over booleans because of extensibility. https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#do-not-use-boolean-fields

Reply:
> But wouldn't this be overkill? I think micro-lvms merged everything back into one image so the only valid array item would be `csi-snapshot`?

Micro-lvms only combines the lvms-node and lvms-controller under a single runtime. The CSI components run separately, but may be deployed within the same pod for simplicity.

Reply:
> We tend to favor string-like fields over booleans because of extensibility. https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#do-not-use-boolean-fields

Thanks! I can never remember where I read that XD

Reply:
> > But wouldn't this be overkill? I think micro-lvms merged everything back into one image so the only valid array item would be `csi-snapshot`?
> 
> Micro-lvms only combines the lvms-node and lvms-controller under a single runtime. The CSI components run separately, but may be deployed within the same pod for simplicity.

Only the csi-snapshot-controller runs externally (and not in the same pod) because its considered an optional part of the control plane. All the other components of CSI run within the lvms controller or node daemon container (not pod):
- csi-external-resizer
- csi-external-snapshotter
- csi-external-provisioner
- csi-node-driver-registrar
- csi-cluster-driver-registrar
- csi-grpc-health-probe

Thats significantly different from most other CSI drivers, and how we achieve a lot of our resource optimization.